### ISSUE TITLE
Another debugging issue...

### ISSUE THREAD

**Issue by Ventajou** [Microsoft/vscode-chrome-debug#42]
Generated a project using this: https://www.npmjs.com/package/scfld-ng-ts
Code is TypeScript, source maps contain the original source. I can start and stop Chome when debugging but breakpoints don't work...

**Comment by roblourens** [comment]
I ran `npm i -g scfld scfld-ng-ts` and `scfld init ng-ts` and got an error - `Could not load the plugin: ng-ts`. When the sourcemaps have the original source inlined, it won't use that source, it will look for the source on disk. But that shouldn't keep it from working.

If you give me some info about the project layout, or the logs with `"diagnosticLogging": true` set, I can probably help you.

**Comment by Ventajou** [comment]
cool, thanks, I can try to do that tonight... is there a way to see the full diagnostic log? the debug window just cuts the lines off if they are too wide

**Comment by roblourens** [comment]
I usually just copy and paste the log into another file for easier reading... I'm sure they'll allow horizontal scrolling or wrapping in the console soon.

**Comment by Ventajou** [comment]
Ok, after looking more carefully at the log, I came across this line:

    SourceMaps.setBreakpoints: c:\code\playground\scfld\app\src\routes.ts can't be resolved to a loaded script.

So I think what may be happening is that my source maps have relative paths for the original files, so the file above is seen as `src/routes.ts`. I tried configuring the Chrome dev tools by adding `c:\code\playground\scfld\app` as a workspace but that didn't seem to do the trick...

**Comment by roblourens** [comment]
Relative paths in sourcemaps should be fine. Check that you have `outDir` set correctly in the launch config. If you're serving files from a server, set `cwd` to the root that files are served from. And you could also try without sourcemaps. And if you still have trouble, post the full logs and I can give you better advice.

**Comment by Ventajou** [comment]
So I have browser-sync serving my files from a folder relative to my project: `.tmp\build` I tried using that value for `outDir` and `cwd` to no avail...

The full log is there https://gist.github.com/Ventajou/bc9908aded4228e43afc

Thanks for the help!

**Comment by roblourens** [comment]
I can't figure it out from the log unfortunately. I have some thoughts on more tracing info to add though... You have a breakpoint in routes.ts, and app.js is loaded later. I assume that's where routes goes. This could be an instance of #43. Does it work if you set the breakpoint after all the scripts are loaded? I'm happy to try myself but I can't get scfld to work.

**Comment by Ventajou** [comment]
I'll try that, thanks. I'd help you with scfld but you probably have better things to do :)

**Comment by coreh** [comment]
Hey there, first of all kudos: This is a really awesome project.

I'm having similar issues as @Ventajou when debugging this sample Webpack project: https://github.com/rauschma/webpack-es6-demo

(I'm _not_ using webpack-dev-server like on issue #40, I'm using a regular, static web server)

I added the following `launch.json` file:

```json
{
	"version": "0.2.0",
	"configurations": [
		{
			"name": "Attach",
			"type": "chrome",
			"request": "attach",
			"port": 9222,
			"diagnosticLogging": true,
			"sourceMaps": true,
			"outDir": "."
		}
	]
}
```

After fiddling a lot with different `outDir` and `cwd` permutations, and modifying `webpack.config.js` to have it generate absolute paths in my source maps, (via `devtoolModuleFilenameTemplate: '[absolute-resource-path]'` in the `output` field) **the debugger now works flawlessly for uncaught Exceptions and for the `debugger;` statement. (yay!)**

On manually set breakpoints, however, it doesn't work, and I still get:

```
SourceMaps.setBreakpoints: /Users/coreh/Projects/webpack-es6-demo/es6/main.js can't be resolved to a loaded script.
```

That happens regardless of whether I set the breakpoints before starting the debugger, or after that.

**Comment by roblourens** [comment]
Hi @coreh, sorry for the slow response. webpack does some weird things - in the sourcemap that it outputs, the `sources` field writes the entries like this - `"webpack:///C:/Users/roblou/code/webpack-es6-demo/es6/main.js"`. The `webpack:///` is strange. It works in Chrome Dev Tools and F12, but I'm not totally sure why. Let me think about how to handle this.

**Comment by roblourens** [comment]
@coreh, this should work if you update to 0.1.9. You will need to do something with ` devtoolModuleFilenameTemplate` to use sourcemaps with webpack, and it should work if you use ` devtoolModuleFilenameTemplate: 'file:///[absolute-resource-path]'`. Note that `outDir` is not needed anymore, and `cwd` is now `webRoot` and is the directory that files are served from, if it's not the directory open in Code. Let me know if you still have issues!

@Ventajou I'm closing this for now, maybe give 0.1.9 a shot and let me know how it goes.

**Comment by mattyork** [comment]
@roblourens this is still not working for me on 0.1.12.

Steps:
* > git clone https://github.com/rauschma/webpack-es6-demo
* > cd webpack-es6-demo
* > npm install
* Change launch.json:

```javascript
 {
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Attach with Sourcemaps",
      "type": "chrome",
      "request": "attach",
      "port": 9222,
      "sourceMaps": true,
      "webRoot": ".",
      "diagnosticLogging": true
    }
  ]
}
```
* Add `devtoolModuleFilenameTemplate: 'file:///[absolute-resource-path]'` to webpack.config.js. New file:

```javascript
var path = require('path');
var webpack = require('webpack');

module.exports = {
    entry: './es6/main.js',
    output: {
        path: __dirname,
        filename: 'bundle.js',
        devtoolModuleFilenameTemplate: 'file:///[absolute-resource-path]'
    },
    module: {
        loaders: [
            {
                loader: 'babel-loader',
                test: path.join(__dirname, 'es6'),
                query: {
                  presets: 'es2015',
                },
            }
        ]
    },
    plugins: [
        // Avoid publishing files when compilation fails
        new webpack.NoErrorsPlugin()
    ],
    stats: {
        // Nice colored output
        colors: true
    },
    // Create Sourcemaps for the bundle
    devtool: 'source-map',
};
```
* > webpack
* > python -m SimpleHTTPServer
* load http://localhost:8000
* Attach debugger
* Set breakpoint in main.js
* Refresh the browser page

Expected: breakpoint is hit
Actua: not hit

This is the error message I get: `SourceMaps.setBP: /Users/mattyork/Dropbox/Projects/webpack-es6-demo/es6/main.js can't be resolved to a loaded script.`

**Comment by roblourens** [comment]
I see the same - I know Edge has an optimization that it will dispose of scripts when they are finished running, if there's no way that script will run again, and I think Chrome probably does the same thing. bundle.js only has code that runs immediately when the page loads, so Chrome can safely get rid of it when it's done. And if I open Chrome devtools on the page, bundle.js isn't there until I refresh the page.

It works for me when I use a launch config instead of attach - basically the same scenario but I guess it attaches quickly enough that Chrome sees that a dev tool is attached and keeps the script. Or, you can refresh the page after attaching, then it will work. A third option is to add some globally accessible function to prevent Chrome from discarding the script, like `window.nocleanupfn = function() { };`.

One problem on Windows - `[absolute-resource-path]` produces a path with mixed \ and / which I need to handle here.

**Comment by sumigoma** [comment]
Was the problem on Windows (mixed \ and /) addressed in the latest release? In chrome debug, my breakpoints seem to point to the appropriate TS files on disk, but breakpoints in VS code still aren't hit. I was wondering if it's because the extension still needs to handle the paths on Windows correctly.

**Comment by Gabriel0402** [comment]
I'm having the same problem. as webpack:///[resourcepath] didn't work, I'm changing the devtoolModuleFilenameTemplate like follows:
```
devtoolModuleFilenameTemplate: function(info){
      return backToForwardSlash("file:///"+info.absoluteResourcePath);
}
```
And there's issues in windows that mixed \ and /, so I add a function to replace all the \ to /, and the path looks correct which is the source files' path. And they're mapping correctly. But in vscode I still cannot set the break point. Neither can I do it in Chrome developer tool. But in chrome after I map the files from file:// to the source files that I added to chrome's workspace, it works. I'm wondering whether the extension does the same thing?  Is there a issue with it? Thanks

**Comment by ChuckDuncan47** [comment]
Since there hasn't been any comments since Dec '15, does that mean it's supposedly working now? 
Unfortunately, I still can't stop at breakpoints.  The app is launched (not attached) and runs nicely.  I am going to try roblauren's suggestion to refresh the page after launching and/or the globally accessible function option.

I will follow this note up with results.  If it fails, I will add as many details as I can.  If anyone currently has more information on this, I'd appreciate hearing about it.  Thanks in advance.

**Comment by Ventajou** [comment]
To be honest I've not tried using this in a long time. Since the Chrome dev tools can't be used when using this plugin then it's not as interesting. Plus the built-in Chrome debugger requires no setup so I've just been using that.

**Comment by roblourens** [comment]
I've fixed a lot of issues in the past few months, so @ChuckDuncan47 please open another issue if you're having problems. You're probably having different problems than the ones in this issue.
